File: PythonInisde/nnCostFunction.py
from __future__ import division
import numpy as np
import sigmoid as s
import scipy.io as sio
import sigmoidGradient as si

def nnCostFunction(nn_params,input_layer_size,hidden_layer_size,num_classes,X,y,lambda1,condition):
    Theta1 = np.reshape(nn_params[0:hidden_layer_size * (input_layer_size + 1)],
    	((input_layer_size + 1),hidden_layer_size));

    Theta2 = np.reshape(nn_params[(hidden_layer_size * (input_layer_size + 1)):],
    ((hidden_layer_size + 1),num_classes));

# Getting back the weights
    Theta1 = np.transpose(Theta1);
    Theta2 = np.transpose(Theta2);
    new_unrolled_Theta = np.concatenate([(np.transpose(Theta1[:,1:])).ravel(),(np.transpose(Theta2[:,1:])).ravel()])
# Number of trainig data sets
    m = X.shape[0];
    Theta1_grad = np.zeros(Theta1.shape);
    Theta2_grad = np.zeros(Theta2.shape);
# Concatenating bias unit
    X = np.concatenate((np.ones((m,1)),X),axis=1)
# Number of features
    n = X.shape[1];

    new_y = np.zeros((num_classes,m))

    for i in range(0,m):
    	new_y[y.item(i),i] = 1;

    first_layer_activation = X;
    second_layer_activation = s.sigmoid(np.dot(Theta1,np.transpose(X)));
# appending bias unit
    second_layer_activation = np.concatenate((np.ones((1,m)),second_layer_activation));
    hypFunction = s.sigmoid(np.dot(Theta2,second_layer_activation));

    first_half = np.sum(np.multiply(new_y,np.log(hypFunction)));
    second_half =np.sum(np.multiply((1-new_y),np.log(1-hypFunction)));

    J = ((-1.0/m)*(first_half+second_half)) + (lambda1/(2*m) *(np.sum(np.multiply(new_unrolled_Theta,new_unrolled_Theta))));
#Calculating delta terms using back propogation
    delta1_grad = np.zeros(Theta1.shape);
    delta2_grad = np.zeros(Theta2.shape);

    for i in range(0,m):
    	a1 = X[i,:][np.newaxis];
    	a2 = s.sigmoid(np.dot(Theta1,np.transpose(a1)));
    	a2 = np.concatenate((np.ones((1,1)),a2));
    	a3 = s.sigmoid(np.dot(Theta2,a2));

    	delta3 = a3 - np.transpose(new_y[:,i][np.newaxis]);
    	delta = np.dot((np.transpose(Theta2)),delta3);
    	delta2 = np.multiply(delta[1:],si.sigmoidGradient(np.dot(Theta1,np.transpose(a1))));

    	delta1_grad = delta1_grad + np.dot(delta2,a1);
    	delta2_grad = delta2_grad + np.dot(delta3,np.transpose(a2));

    Theta1_grad = (1/m) * delta1_grad;
    Theta2_grad = (1/m) * delta2_grad;

    Theta1_grad[:,1:] =  Theta1_grad[:,1:] + (Theta1[:,1:]) * (lambda1/m);
    Theta2_grad[:,1:] = Theta2_grad[:,1:] +  (Theta2[:,1:]) * (lambda1/m);

    grad = np.concatenate([(np.transpose(Theta1_grad)).ravel(),(np.transpose(Theta2_grad)).ravel()]);

    if condition==0:
        return J
    elif condition==1:
        return grad

File: nnCostFunction.m
function [J grad] = nnCostFunction(nn_params, ...
                                   input_layer_size, ...
                                   hidden_layer_size, ...
                                   num_labels, ...
                                   X, y, lambda)
%NNCOSTFUNCTION Implements the neural network cost function for a two layer
%neural network which performs classification
%   [J grad] = NNCOSTFUNCTON(nn_params, hidden_layer_size, num_labels, ...
%   X, y, lambda) computes the cost and gradient of the neural network. The
%   parameters for the neural network are "unrolled" into the vector
%   nn_params and need to be converted back into the weight matrices. 
% 
%   The returned parameter grad should be a "unrolled" vector of the
%   partial derivatives of the neural network.
%

% Reshape nn_params back into the parameters Theta1 and Theta2, the weight matrices
% for our 2 layer neural network
Theta1 = reshape(nn_params(1:hidden_layer_size * (input_layer_size + 1)), ...
                 hidden_layer_size, (input_layer_size + 1));

Theta2 = reshape(nn_params((1 + (hidden_layer_size * (input_layer_size + 1))):end), ...
                 num_labels, (hidden_layer_size + 1));

% Setup some useful variables
m = size(X, 1);
         
% You need to return the following variables correctly 
J = 0;
Theta1_grad = zeros(size(Theta1));
Theta2_grad = zeros(size(Theta2));

% ====================== YOUR CODE HERE ======================
% Instructions: You should complete the code by working through the
%               following parts.
%
% Part 1: Feedforward the neural network and return the cost in the
%         variable J. After implementing Part 1, you can verify that your
%         cost function computation is correct by verifying the cost
%         computed in ex4.m
%
% Part 2: Implement the backpropagation algorithm to compute the gradients
%         Theta1_grad and Theta2_grad. You should return the partial derivatives of
%         the cost function with respect to Theta1 and Theta2 in Theta1_grad and
%         Theta2_grad, respectively. After implementing Part 2, you can check
%         that your implementation is correct by running checkNNGradients
%
%         Note: The vector y passed into the function is a vector of labels
%               containing values from 1..K. You need to map this vector into a 
%               binary vector of 1's and 0's to be used with the neural network
%               cost function.
%
%         Hint: We recommend implementing backpropagation using a for-loop
%               over the training examples if you are implementing it for the 
%               first time.
%
% Part 3: Implement regularization with the cost function and gradients.
%
%         Hint: You can implement this around the code for
%               backpropagation. That is, you can compute the gradients for
%               the regularization separately and then add them to Theta1_grad
%               and Theta2_grad from Part 2.
%

%appending boas unit(ones) to X
X = [ones(m,1) X];

%Number of features
n = size(X,2);

%Converting y's to matrix
new_y = zeros(num_labels,m);
first_half = zeros(num_labels,m);
second_hlaf = zeros(num_labels,m);

for i=1:m,
	new_y(y(i),i)=1;
end;

first_layer_activation = zeros(hidden_layer_size,m);
first_layer_activation = sigmoid(Theta1*X');

%appending bias unit(ones) to first layer activations	
first_layer_activation 	= [ones(1,m) ;first_layer_activation];

hypFunction = sigmoid(Theta2*first_layer_activation);

first_half = sum(sum(new_y.*log(hypFunction)),2);
second_half = sum(sum((1-new_y).*log(1-hypFunction)),2);

% Calculating new unrolled Theta by removing bias units.
new_Theta1 =  Theta1(:,2:n);
new_Theta2 = Theta2 (:,2:(hidden_layer_size+1));
new_unrolled_Theta = [new_Theta1(:); new_Theta2(:)];

J = ((-1/m)*(first_half+second_half)) + (lambda/(2*m) *(sum(new_unrolled_Theta.*new_unrolled_Theta)));

% Calculating partial derivative terms using backpropagation algorithm

delta1_grad = zeros(size(Theta1));
delta2_grad = zeros(size(Theta2));


for t=1:m,
	a1 = X(t,:);
	a2 = sigmoid(Theta1*a1');
	a2 = [ones(1,1); a2];
	a3 = sigmoid(Theta2*a2);

	delta3 = a3 - new_y(:,t);
	delta = Theta2'*delta3; 
	delta2 = (delta(2:end, :)).* sigmoidGradient(Theta1*a1');

	delta1_grad = delta1_grad + delta2 * a1;
	delta2_grad = delta2_grad + delta3 * a2';
end;


Theta1_grad = (1/m) * delta1_grad;
Theta2_grad = (1/m) * delta2_grad;

Theta1_grad(:,2:end) =  Theta1_grad(:,2:end) + (Theta1(:,2:end)) * (lambda/m);
Theta2_grad(:,2:end) = Theta2_grad(:,2:end) +  (Theta2(:,2:end)) * (lambda/m);


% -------------------------------------------------------------

% =========================================================================

% Unroll gradients
grad = [Theta1_grad(:) ; Theta2_grad(:)];


end
